### ISSUE TITLE
Socket data event fires after 5 seconds

### ISSUE THREAD

**Issue by nirvitarka** [nodejs/help#2308]
* **Node.js Version**: 10.16.3
* **OS**: Ubuntu 16.04
* **Scope (install, code, runtime, meta, other?)**: code
* **Module (and version) (if relevant)**: net

I am writing a simple tcp server on ubuntu 16.04. The socket "data" event is fired exactly after 5 seconds.
The same client, when tested with tool "Socket Test" as server, shows no delay in receiving data.

Has anybody faced this issue? Please help.
Here is the code of my server.

In the data event, when I receive a string containing MY_DATA_DELIMITER, I send a COMMAND (just a string not more than 50 characters). The client, as a response to this COMMAND, sends a REPLY. I get this REPLY exactly after 5 seconds.

What am I doing wrong here? 

```js
server.on('connection', function (sock) {
    console.log('CONNECTED: ' + sock.remoteAddress + ':' + sock.remotePort);
    sock.setNoDelay(true);

    sock.on('error', function (err) {
        console.log("ERROR : " + err.toString());
    });

    sock.on('data', function (data) {
        console.log(moment().format("HH:mm:ss") + ' DATA : ' + sock.remoteAddress + ': ' + data);
        data = data.toString().trim();

        if (data.indexOf("MY_DATA_DELIMITER") > -1) {
            console.log(moment().format("HH:mm:ss") + ' sending a command');

            sock.write("MY_COMMAND");
        ////////// I get a response of this write after 5 seconds
        }

    });

    sock.on('close', function (data) {
        console.log('CLOSED: ' + sock.remoteAddress + ' ' + sock.remotePort);
    });

    sock.on('end', function () {
        console.log('ENDED : ' + sock.remoteAddress + ' ' + sock.remotePort);
    });
});
```

**Comment by gireeshpunathil** [comment]
@nirvitarka - here is a re-written version of your code, that works; which also has a sample client code for completeness. I guess the mistake you are doing is, you are setting up the data handler only when the connection is established, by which time the first `data` event would have been already fired and which you did not have an opportunity to capture. How it is firing after 5 minutes is something I am unable to understand; but I guess this should help you to resolve your issue.

#cat 2308.js 

```js
var net = require('net')

var server = net.createServer((sock) => {
  sock.on('data', function (data) {
    console.log(data.toString())
  })
  sock.on('error', function (err) {
    console.log("ERROR : " + err.toString())
  })
  sock.on('close', function (data) {
    console.log('CLOSED: ' + sock.remoteAddress + ' ' + sock.remotePort)
  })
  sock.on('end', function () {
    console.log('ENDED : ' + sock.remoteAddress + ' ' + sock.remotePort)
  })

})
server.on('connection', function (ss) {
  console.log('CONNECTED: ' + ss.remoteAddress + ':' + ss.remotePort)
})

server.listen(12000, () => {
  const client = net.createConnection({ port: 12000 }, () => {
    console.log('connected to server!')
    client.end('world!\r\n')
  })
  client.on('data', (data) => {
    console.log('server data: ' + data.toString())
    client.end()
  })
  client.on('end', () => {
    console.log('disconnected from server')
    server.close()
  })
})
```

```
#node 2308
connected to server!
CONNECTED: <redacted>
world!

ENDED : <redacted>
CLOSED: <redacted>
disconnected from server
```
Hope this helps!

**Comment by nirvitarka** [comment]
@gireeshpunathil, thank you for the reply.
I am just wondering if "data" event can really fire without "connection" event being fired in any case.

Tried the code snippet you sent, it did not help, still 5 seconds delay.

**Comment by gireeshpunathil** [comment]
@nirvitarka - interesting!

can you try `tcpdump` on the code? - `sudo tcpdump -i any -ttt port 12000` on a terminal and run the above in another. The captured trace should show us where the time is lost.

**Comment by nirvitarka** [comment]
I am a newbie to the socket thing... here is the output of tcpdump

sudo tcpdump -i any -ttt port 4444 -c 25
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked), capture size 262144 bytes
 00:00:00.000000 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [P.], seq 33009818:33009858, ack 3327778456, win 2144, length 40
 00:00:00.000060 IP MYSERVERIP.4444 > TCPCLIENTIP.35488: Flags [.], ack 40, win 29200, length 0
 00:00:00.004387 IP MYSERVERIP.4444 > TCPCLIENTIP.35488: Flags [P.], seq 1:24, ack 40, win 29200, length 23
 00:00:00.094573 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [P.], seq 40:42, ack 1, win 2144, length 2
 00:00:00.039879 IP MYSERVERIP.4444 > TCPCLIENTIP.35488: Flags [.], ack 42, win 29200, length 0
 00:00:00.140184 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [.], ack 24, win 2121, length 0
 **00:00:04.829931 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [P.], seq 42:69, ack 24, win 2121, length 27**
 00:00:00.000071 IP MYSERVERIP.4444 > TCPCLIENTIP.35488: Flags [.], ack 69, win 29200, length 0
 00:00:00.116069 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [P.], seq 69:71, ack 24, win 2121, length 2
 00:00:00.000055 IP MYSERVERIP.4444 > TCPCLIENTIP.35488: Flags [.], ack 71, win 29200, length 0
 00:00:15.780793 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [P.], seq 71:98, ack 24, win 2121, length 27
 00:00:00.000093 IP MYSERVERIP.4444 > TCPCLIENTIP.35488: Flags [.], ack 98, win 29200, length 0
 00:00:00.000645 IP MYSERVERIP.4444 > TCPCLIENTIP.35488: Flags [P.], seq 24:47, ack 98, win 29200, length 23
 00:00:00.106288 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [P.], seq 98:100, ack 24, win 2121, length 2
 00:00:00.039877 IP MYSERVERIP.4444 > TCPCLIENTIP.35488: Flags [.], ack 100, win 29200, length 0
 00:00:00.000146 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [.], ack 47, win 2098, length 0
 **00:00:04.952971 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [P.], seq 100:127, ack 47, win 2098, length 27**
 00:00:00.000084 IP MYSERVERIP.4444 > TCPCLIENTIP.35488: Flags [.], ack 127, win 29200, length 0
 00:00:00.099994 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [P.], seq 127:129, ack 47, win 2098, length 2
 00:00:00.000060 IP MYSERVERIP.4444 > TCPCLIENTIP.35488: Flags [.], ack 129, win 29200, length 0

20 packets captured
20 packets received by filter
0 packets dropped by kernel

Can you help spot the issue? This output contains 2 entries when the 5 seconds delay happened, possibly the highlighted ones.

**Comment by gireeshpunathil** [comment]
thanks Sujata. From these 2 lines:

```
00:00:00.140184 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [.], ack 24, win 2121, length 0
00:00:04.829931 IP TCPCLIENTIP.35488 > MYSERVERIP.4444: Flags [P.], seq 42:69, ack 24, win 2121, length 27
```

it appears that the **client** not the server is delaying to write. Let me think over it before coming up with next steps.

**Comment by nirvitarka** [comment]
Moreover, what makes me clueless about the issue is, the same TCPCLIENT has no delay if tested via "nc" command on the same ubuntu server.

**Comment by nirvitarka** [comment]
Turned out to be the way it is handled in CLIENT program. I do not have control over the CLIENT program hence it took much time to debug & figure out what is causing the delay.
The CLIENT expects a line ending with each message as the delimiter. 
Though the example code I shared here uses line ending with message, I missed sending that in my actual code.

Hence the CLIENT kept waiting to detect end of message & this waiting took 5 seconds.

Now when I send the line ending character with each message, it is working without any delay.

Thank you for your time to look at it.
